contracts/vaults/RibbonThetaVault.sol

```diff
@@ -321,76 +318,128 @@ contract RibbonThetaVault is DSMath, OptionsVaultStorage {
-            !depositReceipt.processed
-        ) {
-            _redeemDeposit(currentRound, depositReceipt, false);
-        }
+    /**
+     * @notice Redeems shares that are owed to the account
+     * @param shares is the number of shares to redeem
+     */
+    function redeem(uint256 shares) external nonReentrant {
+        require(shares > 0, "!shares");
+        _redeem(shares, false);
     }
 
-    function redeemDeposit() external nonReentrant {
+    /**
+     * @notice Redeems the entire unredeemedShares balance that is owed to the account
+     */
+    function maxRedeem() external nonReentrant {
+        _redeem(0, true);
+    }
+
+    /**
+     * @notice Redeems shares that are owed to the account
+     * @param shares is the number of shares to redeem, could be 0 when isMax=true
+     * @param isMax is flag for when callers do a max redemption
+     */
+    function _redeem(uint256 shares, bool isMax) internal {
+        require(shares < type(uint104).max, "Overflow");
+
         VaultDeposit.DepositReceipt memory depositReceipt =
             depositReceipts[msg.sender];
-        _redeemDeposit(round, depositReceipt, true);
-    }
 
-    function _redeemDeposit(
-        uint16 currentRound,
-        VaultDeposit.DepositReceipt memory depositReceipt,
-        bool updatesDepositReceipt
-    ) private {
-        require(!depositReceipt.processed, "Processed");
-        require(depositReceipt.round < currentRound, "Round not closed");
-        require(depositReceipt.amount > 0, "!amount");
-
-        uint256 pps = roundPricePerShare[depositReceipt.round];
-        // If this throws, it means that vault's roundPricePerShare[currentRound] has not been set yet
-        // which should never happen.
-        // Has to be larger than 1 because `1` is used in `initRoundPricePerShares` to prevent cold writes.
-        require(pps > PLACEHOLDER_UINT, "Invalid pps");
-
-        // This flag avoid overwriting the processed flag when we are creating
-        // a new deposit receipt
-        // saves 300 gas
-        if (updatesDepositReceipt) {
-            depositReceipts[msg.sender].processed = true;
-        }
+        // This handles the null case when depositReceipt.round = 0
+        // Because we start with round = 1 at `initialize`
+        require(depositReceipt.round < round, "Round not closed");
+
+        uint128 unredeemedShares = _getSharesFromReceipt(depositReceipt);
+
+        shares = isMax ? unredeemedShares : shares;
+        require(shares > 0, "!shares");
+        require(shares <= unredeemedShares, "Exceeds available");
 
-        uint256 shares =
-            uint256(depositReceipt.amount).mul(10**uint256(_decimals)).div(pps);
+        depositReceipts[msg.sender] = VaultDeposit.DepositReceipt({
+            processed: true,
+            round: depositReceipt.round,
+            amount: depositReceipt.amount,
+            unredeemedShares: uint128(uint256(unredeemedShares).sub(shares))
+        });
```

Comment:
wouldn't changing processed -> true and unredeemedShares -> unredeemedShares - x

be more gas effective than reassigning entire struct?

Reply:
Will test this out, from first glance it should be same because the struct is packed into a single storage slot

Reply:
perhaps from a storage perspective it is the same. but I meant in terms of opcodes with reassigning struct as opposed to uints

Reply:
I tested it and your suggestion was better on gas